### PR TITLE
Add `inversion_sigma H`

### PR DESCRIPTION
**Pull request by JasonGross**
Now `inversion_sigma` can be invoked on specific hypotheses, just like
`inversion` and `induction`.

**Kind:** feature

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).
- [ ] Overlay pull requests (if this breaks 3rd party developments in CI, see
https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md for details)

### CODE REVIEW

**Review by herbelin — APPROVED** [review]
A convenient refinement of `inversion_sigma`.

Cc @coq/tactics-maintainers as this is related.

Assigning in the meantime.

File: doc/sphinx/proof-engine/tactics.rst
Comment (jfehrle) [inline doc/sphinx/proof-engine/tactics.rst:2536]:
```suggestion
      When applied to a hypothesis, :tacn:`inversion_sigma` can only handle equalities of the
      listed sigma types.
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (jfehrle) [inline doc/sphinx/proof-engine/tactics.rst:2543]:
```suggestion
      When applied to a hypothesis, :tacn:`inversion_sigma` can only be called on hypotheses that
      are equalities using :g:`Coq.Logic.Init.eq`.
```

"type is syntactically an equality" - won't the user see something like "U = T" in the hypothesis?  Or am I oversimplifying?
Could also say "hypotheses in the form \`U = T\`"

File: theories/Init/Tactics.v
Comment (jfehrle) [inline theories/Init/Tactics.v:326]: If the terms printed in the message are long, this may be hard to read.  Would it be helpful to identify each of terms, e.g. to print `"sig: ", sig` and so forth?  And perhaps putting one on each line with separate idtacs?  Up to you.

**Review by jfehrle — CHANGES_REQUESTED** [review]
Wording suggestions and one or two things to definitely address.

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst:2532]: `sig` and the others are basically literal text, modulo a kludge I'm using to display appropriately qualified names.  (If you pass `-noinit` and then do `Require Coq.Init.Tactics`, then the error message will display appropriately qualified versions of `sig`, etc.  Also similarly if you define `Definition sig := ...`.)  This same kludge inserts spaces before the commas, and I'm not sure how to avoid it while still getting the appropriately qualified versions of the constants in the error messages.

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst:2543]: > "type is syntactically an equality" - won't the user see something like "U = T" in the hypothesis? Or am I oversimplifying?

Yes, that's right.  I'm saying that if you do `Definition foo := @Logic.eq.`, the tactic won't work on things using `foo` rather than `Logic.eq`.  But perhaps "hypotheses of the form :g:`@term = @term`" is fine.  What do you think I should write here?

File: theories/Init/Tactics.v
Comment (JasonGross) [reply theories/Init/Tactics.v:326]: No, absolutely not.  `sig` is going to print as one of `sig`, `Specif.sig`, `Init.Specif.sig`, or `Coq.Init.Specif.sig`, and similarly for the others.  This is the kludge I was talking about above.

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst]: > * change the line `| "inversion_sigma"` to `| "inversion_sigma" ident`

Do you mean `| "inversion_sigma" OPT ident`?

Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst]: Is there some make command to update the rst documentation?  `make doc_gram_rsts` doesn't seem to do it (it doesn't seem to care whether or not I put the `OPT` in, in particular)

Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst]: In any case, I've made the edit and pushed the change

Comment (jfehrle) [reply doc/sphinx/proof-engine/tactics.rst]: > Do you mean `| "inversion_sigma" OPT ident`?

Yes

> Is there some make command to update the rst documentation?

tacn's and cmd's are updated automatically by `make doc_gram_rsts`.  tacv's and cmdv's are not.  Really no good way for a program to figure that out.  We are removing tacv's and cmdv's as we update the documentation.

File: doc/sphinx/proof-engine/tactics.rst
Comment (jfehrle) [reply doc/sphinx/proof-engine/tactics.rst:2532]: > `sig` and the others are basically literal text

I don't follow you.  When I look at the code producing the message, they are variables that are printed.  Can you show me the message from an example?  Maybe then I will get it.

If Coq prints the extra spaces, let's keep them in the message so it's easier to search for in the doc.

File: doc/sphinx/proof-engine/tactics.rst
Comment (jfehrle) [reply doc/sphinx/proof-engine/tactics.rst]: Use :n:, i.e. ``` :n:`@term = @term` ```   :g: doesn't interpret @term.  Or just `u = v`, that should be clear enough, we used that in the equality chapter.

File: doc/changelog/04-tactics/14174-inversion-sigma-on.rst
Comment (jfehrle) [reply doc/changelog/04-tactics/14174-inversion-sigma-on.rst]: How about this part?

> In the description of inversion, the ident seems not to be the name of a hypotheses, so the comparison here seems incorrect.

The "much like" seems a bit of a stretch.  Maybe something like "much like :tacn:`induction` and :tacn:`inversion` when they are used with the name of a hypothesis" would be more accurate?

Comment (JasonGross) [reply doc/changelog/04-tactics/14174-inversion-sigma-on.rst]: I don't think that's necessary; induction and inversion are almost always called with hypotheses, even though they are in fact more general than that.

Comment (jfehrle) [reply doc/changelog/04-tactics/14174-inversion-sigma-on.rst]: ok

**Review by jfehrle — APPROVED** [review]
LGTM

**Comment by herbelin** [comment]
@jfehrle: I assume you're ok with the PR as it is,right?

@JasonGross: For the test-suite, I suspect that it would be easy to add at least an example in `test-suite/success/InversionSigma.v` as a regression test?

**Comment by Zimmi48** [comment]
> @jfehrle: I assume you're ok with the PR as it is,right?

His approval dates from *after* the interaction with Jason, so that's the right way to interpret it.

**Comment by JasonGross** [comment]
> For the test-suite, I suspect that it would be easy to add at least an example in `test-suite/success/InversionSigma.v` as a regression test?

I've done one better.  `inversion_sigma` now supports intropatterns in addition to hypotheses (though it uses `simple_intropattern` rather than `or_and_intropattern_loc`, because the latter is not accessible to `Tactic Notation`; can someone explain to me what the significance is of this distinction, and whether `inversion_sigma` should be made to support `or_and_intropattern_loc`?).  Unfortunately this requires a slight change in the goals left-over (which is technically backwards incompatible but I suspect no one was relying on the exact form), and results in a slow-down of the tactic in some cases, I think because it now generates an intermediate sigma type and calls `destruct` on it, and `destruct` is slow.  I'm not sure if this is worth doing anything about (I'm not sure what can be done about it...)  Now the test-suite tests these variants of `inversion_sigma`.

**Comment by JasonGross** [comment]
Very confused by the metacoq failure
```
File "./theories/Extraction.v", line 31, characters 0-41:
Error:
The informative inductive type sig2 has a Prop instance
in eq_sig2_rec_uncurried (or in its mutual block).
This happens when a sort-polymorphic singleton inductive type
has logical parameters, such as (I,I) : (True * True) : Prop.
The Ocaml extraction cannot handle this situation yet.
Instead, use a sort-monomorphic type such as (True /\ True)
or extract to Haskell.
```
AFAIK, this definition existed already before this PR and was not changed by this PR, nor was it's use changed...

**Comment by herbelin** [comment]
Extraction does not like in the `uncurried` versions that `exist` or `exist2` are applied on equalities, since equalities are in `Prop`, lowering `sig` and `sig2` to `Prop`.

**Comment by JasonGross** [comment]
Ah, I see, thanks, I'll fix.

@Zimmi48 is the new coqbot deployed so that I can ask @coqbot ci minimize
?

File: test-suite/success/InversionSigma.v
Comment (JasonGross) [inline test-suite/success/InversionSigma.v:66]: @ppedrot could you take a look at this and explain what's going on and see if anything can be done about a single call to `induction` taking over 0.5--1.7 seconds?  This is terrible performance, and I don't think my types are *that* big.  (Certainly, I have nested sigma types without primitive projections and this is going to result in large types.  This performance seems a bit egregious, though.)

**Comment by Zimmi48** [comment]
No sorry, not yet.

File: test-suite/success/InversionSigma.v
Comment (JasonGross) [reply test-suite/success/InversionSigma.v:66]: Using primitive projections brings the time down to about 0.05--0.2 seconds, which is better but still not great.
```
total time:      0.256s

 tactic                                   local  total   calls       max
────────────────────────────────────────┴──────┴──────┴───────┴─────────┘
─inversion_sigma_on_as -----------------   0.2%  99.5%       0    0.255s
─induction_sigma_in_as_using -----------   0.2%  96.9%       2    0.248s
─destruct H' as ip ---------------------  67.5%  67.5%       1    0.173s
─induction H as [H'] using (rect _ _ _ _  28.3%  28.3%       1    0.072s
─simpl_ex_proj_in_all ------------------   0.1%   2.5%       1    0.006s
─cbn beta match  delta [ex_proj1 ex_proj   2.4%   2.4%       1    0.006s

 tactic                                   local  total   calls       max
────────────────────────────────────────┴──────┴──────┴───────┴─────────┘
─inversion_sigma_on_as -----------------   0.2%  99.5%       0    0.255s
 ├─induction_sigma_in_as_using ---------   0.2%  96.9%       1    0.248s
 │ ├─destruct H' as ip -----------------  67.5%  67.5%       1    0.173s
 │ └─induction H as [H'] using (rect _ _  28.3%  28.3%       1    0.072s
 └─simpl_ex_proj_in_all ----------------   0.1%   2.5%       1    0.006s
  └cbn beta match  delta [ex_proj1 ex_pr   2.4%   2.4%       1    0.006s
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst:2530]: > I don't follow you. When I look at the code producing the message, they are variables that are printed. Can you show me the message from an example? Maybe then I will get it.

The code is essentially equivalent to `let sig := uconstr:(@sig) in in idtac sig`.  Here is a simpler example with the same principle:
```coq
Module A.
  Definition foo := I.
  Ltac bar := let f := uconstr:(foo) in idtac "foo qualified is" f.
End A.

Goal True.
  A.bar. (* foo qualified is A.foo *)
  Import A.
  bar. (* foo qualified is foo *)
```
Here is an example with the actual tactic:
```coq
Unset Ltac Backtrace.
Goal True = True -> True.
  intro H.
  inversion_sigma H.
  (* Error: Tactic failure: Type of H
is not an equality of recognized Σ types: expected one of sig , sig2 , sigT ,
sigT2 , ex , ex2 but got Prop. *)
```

Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst:2530]: > I don't follow you. When I look at the code producing the message, they are variables that are printed. Can you show me the message from an example? Maybe then I will get it.

The code is essentially equivalent to `let sig := uconstr:(@sig) in in idtac sig`.  Here is a simpler example with the same principle:
```coq
Module A.
  Definition foo := I.
  Ltac bar := let f := uconstr:(foo) in idtac "foo qualified is" f.
End A.

Goal True.
  A.bar. (* foo qualified is A.foo *)
  Import A.
  bar. (* foo qualified is foo *)
```
Here is an example with the actual tactic:
```coq
Unset Ltac Backtrace.
Goal True = True -> True.
  intro H.
  inversion_sigma H.
  (* Error: Tactic failure: Type of H
is not an equality of recognized Σ types: expected one of sig , sig2 , sigT ,
sigT2 , ex , ex2 but got Prop. *)
```

Comment (SkySkimmer) [reply doc/sphinx/proof-engine/tactics.rst:2530]: >This same kludge inserts spaces before the commas, and I'm not sure how to avoid it while still getting the appropriately qualified versions of the constants in the error messages.

Instead of comma-separated you could do space-separated, ie
```suggestion
   .. exn:: Type of @ident is not an equality of recognized Σ types: expected one of sig sig2 sigT sigT2 sigT2 ex or ex2 but got @term
```

Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst:2530]: Do you think that's better?

**Comment by JasonGross** [comment]
I've added a large number of special-cases to reduce the load on unification/`induction`, massively speeding it up.  This also eliminates the need for the simplification tactics (should I remove them entirely?).  @herbelin I think this is ready for re-review and merging.

**Comment by Zimmi48** [comment]
Some squashing is in order. At the very least, the `DO NOT MERGE` commit and its revert.

**Comment by JasonGross** [comment]
I think the do not merge commits might be useful to keep in history?  (I can amend them to remove the do not merge if desired).  I'm happy to do a bit more squashing, though I think it might be useful to keep around the various versions so I can go find them later if, for example, I recall that I had a slow version of this tactic and want to see what version of it was slow.  (I could instead create a branch on my fork that I keep around forever and fully squash all the commits here, if that's preferred?)

**Comment by Zimmi48** [comment]
Yeah, it's usually preferred that the version that is merged has a clean history (no commit amending previous ones in the same PR).

**Comment by JasonGross** [comment]
Rebased and squashed

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst]: Done

**Comment by JasonGross** [comment]
bump @herbelin is there anything else to be done here?

(@Zimmi48 , you said you'd rebase?)

**Comment by Zimmi48** [comment]
Indeed, I'll take care of it.

**Comment by Zimmi48** [comment]
Done.

**Comment by herbelin** [comment]
The PR looks ok to me.

>  (though it uses `simple_intropattern` rather than `or_and_intropattern_loc`, because the latter is not accessible to `Tactic Notation`; can someone explain to me what the significance is of this distinction

The syntax of `or_and_intropattern` is more specific. We could eventually make it available for tactic notations also. So, `simple_intropattern` is too general for `inversion_sigma` but not bad either afaicj.

Merging tomorrow.

**Pull request merged**
